src/mtev_websocket_client.c

```diff
@@ -0,0 +1,461 @@
+  if(res_accept_key == NULL) {
+    mtevL(mtev_error, "Websocket client couldn't find accept key in response headers\n");
+    return mtev_false;
+  }
+  /* skip the header text */
+  res_accept_key += 22; 
+
+  if(!strncmp(accept_key, res_accept_key, mtev_b64_encode_len(SHA_DIGEST_LENGTH))) {
+    return mtev_true;
+  } else {
+    mtevL(mtev_error, "Websocket client found incorrect accept key in response headers\n");
+    return mtev_false;
+  }
+}
+
+static void
+mtev_websocket_client_cleanup(mtev_websocket_client_t *client);
+
+static int
+mtev_websocket_client_drive(eventer_t e, int mask, void *closure, struct timeval *now) {
+  mtev_websocket_client_t *client = closure;
+
+  if(mask & EVENTER_EXCEPTION || client->should_close || client->closed) {
+abort_drive:
+    mtev_websocket_client_cleanup(client);
+    return 0;
+  }
+
+  if(!client->did_handshake) {
+    if(!client->sent_handshake && mask & EVENTER_WRITE) {
+      if(mtev_websocket_client_send_handshake(client) == mtev_false) {
+        mtevL(mtev_error, "mtev_websocket_client_send_handshake failed, aborting drive\n");
+        goto abort_drive;
+      }
+      client->sent_handshake = mtev_true;
+    }
+    
+    if(client->sent_handshake && mask & EVENTER_READ) {
+      if(mtev_websocket_client_recv_handshake(client) == mtev_false) {
+        mtevL(mtev_error, "mtev_websocket_client_recv_handshake failed, aborting drive\n");
+        goto abort_drive;
+      }
+      wslay_event_context_client_init(&client->wslay_ctx, &wslay_callbacks, client);
+      client->did_handshake = mtev_true;
+    } else {
+      return EVENTER_READ | EVENTER_EXCEPTION;
+    }
+  }
+
+  if (wslay_event_want_read(client->wslay_ctx) == 0
+      && wslay_event_want_write(client->wslay_ctx) == 0) {
+    mtevL(mtev_error, "Websocket client's wslay context didn't want read or write, aborting drive\n");
+    goto abort_drive;
+  }
+
+  /* wslay_on_msg_recv_callback may set client->should_close if the consumer handler fails */
+  if (wslay_event_recv(client->wslay_ctx) != 0 || client->should_close) {
+    mtevL(mtev_error, "Websocket client's wslay_event_recv failed, aborting drive\n");
+    goto abort_drive;
+  }
+
+  if (wslay_event_send(client->wslay_ctx) != 0) {
+    mtevL(mtev_error, "Websocket client's wslay_event_send failed, aborting drive\n");
+    goto abort_drive;
+  }
+
+  return client->wanted_eventer_mask | EVENTER_EXCEPTION | EVENTER_WRITE;
+  return 0;
+}
+#endif
+
+mtev_websocket_client_t *
+mtev_websocket_client_new(const char *host, int port, const char *path, const char *service,
+                          websocket_client_msg_handler callback) {
+#ifdef HAVE_WSLAY
+  int fd = -1, rv;
+  struct addrinfo *res = NULL, *res_ptr = NULL, hints;
+  char port_str[6];
+
+  snprintf(port_str, 6, "%d", port);
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  if((rv = getaddrinfo(host, port_str, &hints, &res))) {
```

Comment:
This is a blocking call. I would have this function just take an IPv4 or IPv6 address -- no lookup.  The caller can struggle with the non-blocking lookup.